barretenberg/cpp/src/barretenberg/commitment_schemes/ipa/ipa.hpp

```diff
@@ -85,72 +175,94 @@ template <typename Curve> class IPA {
             /*finite_field_multiplications_per_iteration=*/1);
 
         // Iterate for log(poly_degree) rounds to compute the round commitments.
-        auto log_poly_degree = static_cast<size_t>(numeric::get_msb(poly_degree));
-        std::vector<GroupElement> L_elements(log_poly_degree);
-        std::vector<GroupElement> R_elements(log_poly_degree);
-        std::size_t round_size = poly_degree;
-
-        // Allocate vectors for parallel storage of partial products
-        const size_t num_cpus = get_num_cpus();
-        std::vector<Fr> partial_inner_prod_L(num_cpus);
-        std::vector<Fr> partial_inner_prod_R(num_cpus);
-        // Perform IPA rounds
+        auto log_poly_degree = static_cast<size_t>(numeric::get_msb(poly_degree_plus_1));
+
+        // Allocate space for L_i and R_i elements
+        GroupElement L_i;
+        GroupElement R_i;
+        std::size_t round_size = poly_degree_plus_1;
+
+#ifndef NO_MULTITHREADING
+        //  The inner products we'll be computing in parallel need a mutex to be thread-safe during the last
+        //  accumulation
+        std::mutex inner_product_accumulation_mutex;
+#endif
+        // Step 6.
+        // Perform IPA reduction rounds
         for (size_t i = 0; i < log_poly_degree; i++) {
```

Comment:
could flip this backwards to align with documentation

Reply:
Forgot to do that. One of the reasons I got rid of the vectors

Reply:
Actually, the index in the manifest is correct. I'm just iterating like this for simplicity, so I'd prefer not to change it